### ISSUE TITLE
Cannot load batchnorm layer of TensorFlow(tf.contrib.layers.batch_norm)

### ISSUE THREAD

**Issue by stereomatchingkiss** [opencv/opencv#9660]
python3.5
TensorFlow 1.3
Ubuntu 16.04.1

Minimum codes:

```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import numpy as np
import tensorflow as tf

features = tf.placeholder(tf.float32, [None, 128, 128, 2], name = 'input')
model = tf.reshape(features, [-1, 128*128*2])
model = tf.contrib.layers.batch_norm(model, is_training = False)
model = tf.layers.dense(inputs = model, units = 8, name = 'regression_output')

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    saver = tf.train.Saver()
    saver.save(sess, 'reshape.ckpt')
    tf.train.write_graph(sess.graph.as_graph_def(), "", 'graph.pb')
```

Freeze the model:

python3 ~/.keras2/lib/python3.5/site-packages/tensorflow/python/tools/freeze_graph.py --input_graph=graph.pb --input_checkpoint=reshape.ckpt --output_graph=frozen_graph.pb --output_node_names=regression_output/BiasAdd

Load the model:

```
#include <opencv2/core.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/highgui.hpp>

#include <iostream>

using namespace cv;

int main()try
{
    std::string const model("/home/whatever/deep_homography/cnn/tensorflow/frozen_graph.pb");

    dnn::Net net = dnn::readNetFromTensorflow(model);
    if(net.empty()){
        std::cerr<<"Can't load network by using the mode file:"<<std::endl;
        std::cerr<<model<<std::endl;
        return -1;
    }

    return 0;
}catch(std::exception const &ex){
    std::cerr<<ex.what()<<std::endl;
}

```

Error message:

OpenCV Error: Unspecified error (More than one input is Const op) in getConstBlob, file /home/whatever/opencv/modules/dnn/src/tensorflow/tf_importer.cpp, line 513
/home/whatever/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:513: error: (-2) More than one input is Const op in function getConstBlob

**Comment by dkurt** [comment]
@stereomatchingkiss, try to fuse constant ops by
```bash
~/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
  --in_graph=frozen_graph.pb \
  --out_graph=fused_graph.pb \
  --inputs=input_node_name \
  --outputs=output_node_name \
  --transforms="fold_constants sort_by_execution_order"
```

**Comment by stereomatchingkiss** [comment]
@dkurt Tried it, still pop out same error message

OpenCV Error: Unspecified error (More than one input is Const op) in getConstBlob, file /home/whatever/opencv/modules/dnn/src/tensorflow/tf_importer.cpp, line 513
/home/whatever/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:513: error: (-2) More than one input is Const op in function getConstBlob

**Comment by dkurt** [comment]
@stereomatchingkiss, let me ask a question about the model. Did you want to apply a batch normalization over the input image or not? If I'm correct and the model is `input->bn->reshape->fc` the following sample works for the latest state of OpenCV:

```python
features = tf.placeholder(tf.float32, [None, 128, 128, 2], name = 'input')
model = tf.contrib.layers.batch_norm(features, is_training = False)
model = tf.reshape(model, [-1, 128*128*2])
model = tf.layers.dense(inputs = model, units = 8, name = 'regression_output')
```
Then save a checkpoint and the graph definition.
```bash
python ~/tensorflow/tensorflow/python/tools/freeze_graph.py \
  --input_graph=graph.pb \
  --input_checkpoint=reshape.ckpt \
  --output_graph=frozen_graph.pb \
  --output_node_names=regression_output/BiasAdd

python ~/tensorflow/tensorflow/python/tools/optimize_for_inference.py \
  --input frozen_graph.pb \
  --output opt_graph.pb \
  --frozen_graph True \
  --input_names input \
  --output_names regression_output/BiasAdd

~/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
  --in_graph=opt_graph.pb \
  --out_graph=fused_graph.pb \
  --inputs=input \
  --outputs=regression_output/BiasAdd \
  --transforms="fold_constants sort_by_execution_order"
```
Launch:
```python
import cv2 as cv
import numpy as np

net = cv.dnn.readNetFromTensorflow('fused_graph.pb')
inp = np.random.standard_normal([1, 2, 128, 128]).astype(np.float32)
net.setInput(inp)
out = net.forward()
```

**Comment by stereomatchingkiss** [comment]
@dkurt Thanks, after I pull latest codes from master branch and rebuild, it works now. I would suggest someone write down how to deploy tensorflow network with batchnorm on wiki, I do not have the right to access the wiki, else I will do it myself.

rough example:

```
1. clone https://github.com/tensorflow/tensorflow
2. cd to tensorflow
3. install bazel if you haven't
4. enter "/.configure"
5. enter "bazel build tensorflow/tools/graph_transforms:transform_graph"
6. freeze graph

python ~/tensorflow/tensorflow/python/tools/freeze_graph.py \
  --input_graph=graph.pb \
  --input_checkpoint=reshape.ckpt \
  --output_graph=frozen_graph.pb \
  --output_node_names=regression_output/BiasAdd

7. optimize graph

python ~/tensorflow/tensorflow/python/tools/optimize_for_inference.py \
  --input frozen_graph.pb \
  --output opt_graph.pb \
  --frozen_graph True \
  --input_names input \
  --output_names regression_output/BiasAdd

8. transform graph

~/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
  --in_graph=opt_graph.pb \
  --out_graph=fused_graph.pb \
  --inputs=input \
  --outputs=regression_output/BiasAdd \
  --transforms="fold_constants sort_by_execution_order"
```

**Comment by stereomatchingkiss** [comment]
@dkurt Latest solution only work if the batchNorm layer is always set as False by constant value, you cannot set it by Variable or placeholder

    ```
input_shape = [None,128,128,2]
    features = tf.placeholder(tf.float32, input_shape, name = 'input')
    net = tf.layers.conv2d(inputs = features, filters = 64, kernel_size = [3, 3], strides = (2, 2), padding = 'same')
    #training = tf.placeholder(tf.bool, name='training')
    training = tf.Variable(False, name='training', trainable = False)
    net = tf.contrib.layers.batch_norm(net, is_training = training)
    net = tf.nn.relu(net)
    net = tf.reshape(net, [-1, 64 * 64 * 64])
    net = tf.layers.dense(inputs = net, units = 8, kernel_initializer = tf.contrib.layers.xavier_initializer(), name = 'regression_output')

    global_step = tf.Variable(0, trainable=False)
    boundaries = [30000, 60000]
    values = [0.005, 0.0005, 0.00005]
    learning_rate = tf.train.piecewise_constant(global_step, boundaries, values)

    x = tf.placeholder(tf.float32, shape=[None,8])
    #apply sqrt to avoid the loss value bloated
    loss = tf.sqrt((tf.nn.l2_loss(net - x)))

    with tf.Session() as sess:
        sess.run(tf.global_variables_initializer())
        imgs = np.zeros([1,128,128,2])
        labels = [[1,2,3,4,5,6,7,8]]
        #change training to False before save
        sess.run(loss, feed_dict = {features : imgs, training : False, x : labels})
        
        saver = tf.train.Saver()
        saver.save(sess, 'reshape_final.ckpt')        
        tf.train.write_graph(sess.graph.as_graph_def(), "", 'graph_final.pb')
```

After that, freeze the graph

`python3 ~/.keras2/lib/python3.5/site-packages/tensorflow/python/tools/freeze_graph.py --input_graph=graph_final.pb --input_checkpoint=reshape_final.ckpt --output_graph=frozen_graph.pb --output_node_names=regression_output/BiasAdd`

optimize

`python3 ~/.keras2/lib/python3.5/site-packages/tensorflow/python/tools/optimize_for_inference.py --input frozen_graph.pb --output opt_graph.pb --frozen_graph True --input_names input --output_names regression_output/BiasAdd`

transform

`~/Qt/3rdLibs/tensorflow/bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=opt_graph.pb --out_graph=fused_graph.pb --inputs=input --outputs=regression_output/BiasAdd --transforms="fold_batch_norms fold_old_batch_norms fold_constants sort_by_execution_order"
`

Load by c++

    ```
std::string const model("/home/ramsus/Qt/blogCodes2/deep_homography/cnn/tensorflow /fused_graph.pb");

    dnn::Net net = dnn::readNetFromTensorflow(model);
    if(net.empty()){
        std::cerr<<"Can't load network by using the mode file:"<<std::endl;
        std::cerr<<model<<std::endl;
        throw std::runtime_error("net is empty");
    }
```

Get error messages:

> BatchNorm/moments/mean:Mean(conv2d/convolution)(BatchNorm/moments/mean/reduction_indices)
> keep_dims:[ ]
> Tidx:[ ]
> T:0
> OpenCV Error: Unspecified error (Unknown layer type Mean in op BatchNorm/moments/mean) in populateNet, file /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/tensorflow/tf_importer.cpp, line 1077
> /home/ramsus/Qt/3rdLibs/opencv/modules/dnn/src/tensorflow/tf_importer.cpp:1077: error: (-2) Unknown layer type Mean in op BatchNorm/moments/mean in function populateNet
>

**Comment by mhaghighat** [comment]
I don't know why this issue is closed. The problem still exists in OpenCV 3.4.1.

**Comment by dkurt** [comment]
@mhaghighat, please provide a minimal example to reproduce the problem.
